Mitigate M7D drone's inability to repair huge stations

jamesjonesphoenix opened the pull request:
M7D repair drones seem to be incapable of repairing a shipyard in-sector. They buzz around but never fire their repair lasers. They seem to stay about 4km out from the centre of the station due to the large bounding box of the model and with a repair laser range of 2.2km they never get around to firing. 

The solution I've coded up consists of two parts:
1. Switching off collision for the drones if the repair target size exceeds a certain value (and no enemies are detected nearby). If I only do this, the drones will fly one effective attack run and then fail to perform another as they buzz around within 200m of the centre of the station, failing to line up more shots ever again. Hence part 2:
2. If the drone finds itself less than 300m from the centre of the station, I execute `= [THIS]-> move to position: x=$xPos y=$yPos z=$zPos with precision 100 m` to a position 2700m away from the station in a random direction. This allows them to perform another effective attack run, and then the cycle repeats. 

What do you think of this approach? I've tested (with Argon Shipyard only) and it actually works quite well. It looks slightly silly, but I should be able to mitigate that with some tweaking of the numbers. Similar to https://github.com/Cycrowuk/X3FLPatch/pull/3, I'm not sure if `move to position continue` or `move to position` is best for this use case.

Cycrowuk commented:
It might be possible to use the docking command (althouh this will only work if the ship can actually dock).  Then when it gets close enough, stop the command and fire the laser while remaining stationary. 

Docking is the only time a ship will enter the collision box of a station.

The advantage here is that the docking line is kept clear, so it can ignore collisions to the station and fly towards it, without worry about actually crashing into the structure.

You do that elsewhere and you run the risk of it crashing into the side of the station.

if i get time i may look into adding a new attack run command that can line up with the closest part of an object (instead of the centre)

jamesjonesphoenix commented:
I've been trying to come up with a new idea for this problem. I tried ` [THIS]-> fire laser of turret null ahead` but it doesn't seem to do anything and I can't find an example of it being used in other scripts to troubleshoot.

Other than that, my next idea is to create a target between the station and the drone. My idea would work like this:
  - Grab position of drone and pos of station at the moment it aborts docking and create a vector between them. 
  - Calculate a point along the line of that vector about 1500m from the drone.
  - Create an invisible special object at this point and call `$drone-> fire lasers on target $target using turret null` on this point. 
  - Station isn't in range, but the invisible special obj will be and since the drone is near the docking clamp it should hit the station when firing at the new target.

Does this sound feasible to you? Particularly I'm hoping you can tell me whether a ship will shoot at an invisible special object before I go to the trouble of re-learning how to do trigonometry.

jamesjonesphoenix commented:
Thanks for the `fire ahead` command. I've got it all working pretty well now. Brace yourself, `!repair.repairobject` is almost 750 lines. Here's the basic logic flow:

  - Try docking method first
    - Ship attempts to dock. If impossible or no free docking ports, approach the station manually.
	- Abort docking once within 700m of a vertex
    - After aborting docking issue `move to position` command onto drone's current position with collision switched off. This stops the ship without triggering collision avoidance.
	- Move laterally to move out of way of other approaching drones and rotate to somewhat random angle.
	  - I added `!lib.rotateobject.smooth` secondary script to rotate an object smoothly. Can be called from `!lib.rotateobject` if `a.smooth` arg is set.
	    - Also corrected offset calc so it never produces an angle greater than 65536 or less than 0.
    - Drone fires.
	  - Collisions switched back on while firing.
	  - Try firing ahead first.
	  - If that fails try firing at target.
	  - For each failure we call `[THIS]-> set ship speed to $speed` to thrust closer to station centre. 
      - `[SIGNAL_REPAIRED]` sets a local var on the drone to mark the last time a laser hit. 
      - If laser isn't hitting, rotate toward station centre and call `[THIS]-> set ship speed to $speed`. Very handy because it does not set off collision avoidance. Try firing again. 
      - Check distance to nearest vertex. If < 150m perform an escape maneuver and start all over again.
	    - Collision switched off during escape maneuver. Drones love suiciding into walls, so this is most likely time to see drone clipping into station. 
  - If docking is impossible we approach the station manually. This also works well, but has small chance of drone clipping the station.
    - Maneuver close to the station with `movetoposition`. Ignore collisions while doing this. 
	  - Regularly check if a docking port becomes free. If yes, then switch to docking method. 
	- If `movetoposition` finished and still not close enough, rotate drone towards station centre then repeatedly call `[THIS]-> set ship speed to $speed` until within 600m of a vertex.
    - Fire lasers as per docking method. 	
  - When repairs are complete we return to carrier. Unfortunately the drones loved to suicide into the station so I have disabled collisions while close to the station in `!repair.repairobject.check`. This causes a high chance of clipping but better than the alternative.  

I've included code to mitigate issue [7](https://github.com/Cycrowuk/X3FLPatch/issues/7). This switches drone's race to `[Other]` if necessary before firing lasers, and back afterwards. It all seems to work pretty well, though there may be edge cases I haven't accounted for. We set IFF as per original race and store old race in a local var.

I'm using docking method and manual approach methods for any stationary object, as they usually work better than attack runs. 

I've tested the following stations with docking disabled and enabled: 
  - Big Stations
    - Terran Shipyard 
    - Aldrin Ship Production Base - (bit of a schlep because 1. centre of the station is out in space and 2. Some vertexes of the asteroid part don't seem to register. The drones still got the job done.)
    - Terran Military Base
    - Military Outpost
    - USC Orbital Support Station
    - Atreus Shipyard
	- Atreus HQ
	- Argon Shipyard
  - Small Stations
    - Quantum Tube Fab
	- Chip Plant
	- Ammunition Factory
	- Silicon Mine
	- Crystal Fab
	- Weapon Component Factory
	- Bofu Chem Lab
	- Plankton Farm
	- Drone Production Facility
	- Boron Trading Station
	- Recon Drone Construction Facility
	- Ore Mine
	- Paranid Equipment Dock
	- Boarding Pod Forge
  - Large Orbital Weapons Platform
  - If there's a particular station you think could cause trouble, let me know and I'll test it. 
 
Questions:
  - Have a look, let me know any changes you think are needed. Here's my object logging script for if you want to test commit in its current state - [plugin.jj.lib.log.object.xml](https://github.com/user-attachments/files/23394012/plugin.jj.lib.log.object.xml)
  - Performance takes a hit when 20 drones are continously firing. This will improve somewhat when I remove logging (Current commit is lousy with debug statements.). But if you notice any opportunities to reduce CPU load please let me know. In particular: 
	- What are `sub.RemoveLasers:` and `sub.RestoreLasers:` necessary for? They are called every time we fire in original script and I've kept them without knowing why.
	- Is using [SIGNAL_REPAIRED] like this CPU intensive? If yes, is there a way to mitigate it?
	- Would you consider reducing repair laser fire rate and increasing repair per shot? 
	  - Is there any way to get rate of fire in scripting engine? I was trying to get this to change wait times in the firing loops. Not sure how much this would help though. 
  - Does this script need to be tested for any other type of repair ship?
  - We check whether repairer is a `[Big Ship]` in the original script but it wasn't used for anything. I removed it - do you know the intention for this?
  
TODO which I'll do as part of PR [M7D Improvements](https://github.com/Cycrowuk/X3FLPatch/pull/6): 
  - I want to use these methods on stationary moveable ships too. I'll add this later by checking for running `!move.waitforrepair` script to flag the ship as 'stationary'.

jamesjonesphoenix commented:
Added `!move.waitforrepair` script to interrupt target's current task and stop until repaired. Works similarly to `!move.waitfordocked` in that an array of repairers is stored in a local var in the target ship and checked in a loop. 

Called from `!repair.repairobject` which repairs a stopped ship as if it was a stationary object. Loop checks if ship has started moving and goes back to standard attack runs if so.

This only runs on non-player owned target ships. I'll consider running it for player ships later, but 

I found that targets could easily escape from approaching M7D before drone launch so I've written functionality to make target also wait for them in `!move.waitforrepair`, even though they are not repair ships in the literal sense. M7D are stored in a separate local var with separate checks. I add M7Ds in `!ship.cmd.repairobject.drones` and call `!move.waitforrepair` from there. This is working well but I'll commit `!ship.cmd.repairobject.drones` to [PR 6 - M7D Improvements ](https://github.com/Cycrowuk/X3FLPatch/pull/6) instead of this one to avoid Git conflicts.

Note these files are still full of debug statements. Once reviewed I'll remove them.
